File: src/leetcode/python/PopulatingNextRightPointersInEachNode_116.py
class TreeLinkNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None


class PopulatingNextRightPointersInEachNode_116(object):
    def connect(self, root):
        """
         :type root: TreeLinkNode
         :rtype: nothing
         """
        def node_connect(left, right):
            if not left: return
            left.next = right
            right.next = None
            node_connect(left.left, left.right)
            node_connect(left.right, right.left)
            node_connect(right.left, right.right)

        if not root: return
        root.next = None
        node_connect(root.left, root.right)

File: src/leetcode/java/PopulatingNextRightPointersInEachNode_116.java
package leetcode.java;

public class PopulatingNextRightPointersInEachNode_116 {
  public void connect(TreeLinkNode root) {
    if (root == null) return;
    root.next = null;
    nodeConnect(root.left, root.right);
  }

  public static void nodeConnect(TreeLinkNode left, TreeLinkNode right) {
    if (left == null) return;
    left.next = right;
    right.next = null;
    nodeConnect(left.left, left.right);
    nodeConnect(left.right, right.left);
    nodeConnect(right.left, right.right);
  }

  class TreeLinkNode {
    int val;
    TreeLinkNode left, right, next;
    TreeLinkNode(int x) { val = x; }
  }
}
